src/commonMain/kotlin/org/cru/godtools/tool/model/Multiselect.kt

```diff
@@ -0,0 +1,99 @@
+private const val XML_OPTION_VALUE = "value"
+
+class Multiselect : Content {
+    internal companion object {
+        internal const val XML_MULTISELECT = "multiselect"
+    }
+
+    @VisibleForTesting
+    internal val stateName: String
+    @VisibleForTesting
+    internal val selectionLimit: Int
+
+    val options: List<Option>
+
+    internal constructor(parent: Base, parser: XmlPullParser) : super(parent, parser) {
+        parser.require(XmlPullParser.START_TAG, XMLNS_CONTENT, XML_MULTISELECT)
+
+        stateName = parser.getAttributeValue(XML_STATE).orEmpty()
+        selectionLimit = (parser.getAttributeValue(XML_SELECTION_LIMIT)?.toIntOrNull() ?: 1).coerceAtLeast(1)
+
+        options = mutableListOf()
+        parser.parseChildren {
+            when (parser.namespace) {
+                XMLNS_CONTENT -> when (parser.name) {
+                    XML_OPTION -> options += Option(this, parser)
+                }
+            }
+        }
+    }
+
+    @RestrictTo(RestrictTo.Scope.TESTS)
+    internal constructor(
+        parent: Base = Manifest(),
+        stateName: String = "",
+        selectionLimit: Int = 1,
+        options: ((Multiselect) -> List<Option>)? = null
+    ) : super(parent) {
+        this.stateName = stateName
+        this.selectionLimit = selectionLimit
+        this.options = options?.invoke(this).orEmpty()
+    }
+
+    override val isIgnored get() = FEATURE_MULTISELECT !in ParserConfig.supportedFeatures || super.isIgnored
+
+    class Option : Content, Parent {
+        private val multiselect: Multiselect
+
+        @VisibleForTesting
+        internal val value: String
+
+        override val content: List<Content>
+
+        internal constructor(multiselect: Multiselect, parser: XmlPullParser) : super(multiselect, parser) {
+            this.multiselect = multiselect
+            parser.require(XmlPullParser.START_TAG, XMLNS_CONTENT, XML_OPTION)
+
+            value = parser.getAttributeValue(XML_OPTION_VALUE).orEmpty()
+
+            content = parseContent(parser)
+        }
+
+        @RestrictTo(RestrictTo.Scope.TESTS)
+        internal constructor(multiselect: Multiselect, value: String = "") : super(multiselect) {
+            this.multiselect = multiselect
+            this.value = value
+            content = emptyList()
+        }
+
+        fun isSelected(state: State) = value in state.getAll(multiselect.stateName)
+        fun isSelectedFlow(state: State) = state.changeFlow(multiselect.stateName)
+            .map { isSelected(state) }
+            .distinctUntilChanged()
+        fun toggleSelected(state: State): Boolean {
+            val current = state.getAll(multiselect.stateName)
+            when {
+                value in current -> state.removeValue(multiselect.stateName, value)
+                current.size < multiselect.selectionLimit -> state.addValue(multiselect.stateName, value)
+                multiselect.selectionLimit == 1 -> state[multiselect.stateName] = value
```

Comment:
So this does three things

1. When a value is selected and it already exists in the state, remove it.
2. If a value is selected, is not already apart of the state, and the current size of the state is less then the selection limit, add it to the state
3. If the selectionLimit is only one, set the state to be equal to that selected value.

Is this correct?

Reply:
Also if it hits say the first line ```value in current...``` does it stop in the when block? I assume it would have to because if not it would remove a selected value and then just readd it.

Reply:
that is correct, I should have written tests for all the behaviors.

Basically If `selectionLimit` is 1 I want the options to behave like radio buttons, where the last one you select overrides any other selection. Otherwise I want to just observe the selection limit and not allow the user to select any beyond the selection limit

Reply:
when blocks in Kotlin do not support fall-through